optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/drools/DroolsScoreDirector.java

```diff
@@ -75,15 +79,64 @@ private void resetKieSession() {
         }
         kieSession = scoreDirectorFactory.newKieSession();
         ((RuleEventManager) kieSession).addEventListener(new OptaplannerRuleEventListener());
-        workingScoreHolder = getScoreDefinition().buildScoreHolder(constraintMatchEnabledPreference);
-        kieSession.setGlobal(GLOBAL_SCORE_HOLDER_KEY, workingScoreHolder);
+        resetWorkingScoreHolder();
         // TODO Adjust when uninitialized entities from getWorkingFacts get added automatically too (and call afterEntityAdded)
         Collection<Object> workingFacts = getWorkingFacts();
         for (Object fact : workingFacts) {
             kieSession.insert(fact);
         }
     }
 
+    private void resetWorkingScoreHolder() {
+        workingScoreHolder = getScoreDefinition().buildScoreHolder(constraintMatchEnabledPreference);
+        scoreDirectorFactory.getRuleToConstraintWeightExtractorMap().forEach(
+                (Rule rule, Function<Solution_, Score> extractor) -> {
+            Score constraintWeight = extractor.apply(workingSolution);
+            if (constraintWeight.getInitScore() != 0) {
+                Class<?> constraintConfigurationClass = getSolutionDescriptor().getConstraintConfigurationDescriptor()
+                        .getConstraintConfigurationClass();
+                throw new IllegalArgumentException("The constraintWeight (" + constraintWeight
+                        + ") for constraintPackage (" + rule.getPackageName()
+                        + ") and constraintName (" + rule.getName()
+                        + ") of constraintConfigurationClass (" + constraintConfigurationClass
+                        + ") must have an initScore (" + constraintWeight.getInitScore() + ") equal to 0.\n"
+                        + "Maybe validate your " + constraintConfigurationClass.getSimpleName() + " data input.");
+            }
+            if (!getScoreDefinition().isPositive(constraintWeight)) {
+                Class<?> constraintConfigurationClass = getSolutionDescriptor().getConstraintConfigurationDescriptor()
+                        .getConstraintConfigurationClass();
+                throw new IllegalArgumentException("The constraintWeight (" + constraintWeight
+                        + ") for constraintPackage (" + rule.getPackageName()
+                        + ") and constraintName (" + rule.getName()
+                        + ") of constraintConfigurationClass (" + constraintConfigurationClass
+                        + ") must have a positive or zero constraintWeight (" + constraintWeight + ").\n"
```

Comment:
Duplicate occurrence of the `constraintWeight` parameter. I think it would be sufficient and more natural to say "The constraintWeight (-1) ... must be positive or zero."

Reply:
I prefer to have the -1 very close to the "must be positive or zero", so the reason is more obvious. But you're right, the duplication here is unneeded, changing it to:

```

                throw new IllegalArgumentException("The @" + ConstraintWeight.class.getSimpleName()
                        + " annotation for constraintPackage (" + rule.getPackageName()
                        + ") and constraintName (" + rule.getName()
                        + ") of constraintConfigurationClass (" + constraintConfigurationClass
                        + ") must have a positive or zero constraintWeight (" + constraintWeight + ").\n"
                        + "Maybe validate your " + constraintConfigurationClass.getSimpleName() + " data input.");
```